workflow/commit-messages.md

```diff
@@ -0,0 +1,43 @@
+---
+title: Commit guidelines
+description: Some general guidelines to help you in writing good commits for collaboration and clarity.
+---
+
+# Commit Guidelines
+
+As a general guideline you can follow the idea that _a single commit should enclose a single problem/feature that has been fixed/added_.
+
+This way of organizing your changes to the codebase makes is easy for reviewers and people who try to understand the history of a project to determine what has been happening in the project.
+
+Here a both a good and a bad example of of commit message.
+
+**Bad:**
+
+```
+fixes bug
+```
+
+**Good:**
+
+```
+Fix bug where database gets deleted on POST request
+```
+
+## Breaking up large tickets
+
+If your feature/ticket includes mutliple things that you have to complete in order to submit a pull request, these should all be seperate commit messages, according to the general guideline above. This will make it easy for the reviewer to understand what exactly has happened.
+
+For adding a new page in an example Project this could look like the following
+
+```
+Add the Person model
+Configure the Views to use the new Person Model
+Use the Data from the Views in the index template
+```
+
+For more information about how to write good commit messages you can check out [this great article by Chris Beams](https://chris.beams.io/posts/git-commit/)
+
+## Rebasing
+
+If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.
```

Comment:
Although I'm not sure if there should be "a team standard" (there may be project standards), but personally I much prefer the GitHub default of "Merge commit", which uses a standard "git merge" to merge a feature branch with master.

Using "squash" loses a significant amount of information about how the work had been split up. We are supposed to try to keep PRs as small as possible, but there are definitely mixed results with this, and even if you do this well there are cases where adding one whole piece of functionality requires changes in quite a number of places.

If someone has taken the time to separate out the creation of the "models.User" from "views.login" as 2 separate commits (as I'd hope they might), I don't want to lose that separation by squashing them together.

This does somewhat depend on how good we are at keeping our commits neat. Obviously I'd rather have 1 squashed commit for the whole PR than 30 commits called "fixed typo". But for my PRs, I'd certainly rather the history was preserved as I wrote it.

Reply:
If we are talking about GitHub 'Squash on merge' feature it simply squashes whole PR into one commit and rebases it on top of master (or whatever branch you are merging to). As far as I remember (we use it by default in [build.snapcraft.io](https://github.com/canonical-websites/build.snapcraft.io)) it creates a commit message that lists all the messages of commits being squashed.

I believe we had a dev catch up discussion about merging vs rebasing vs squashing PRs and the consensus was to use merge by default not to loose commit history by squashing whole PR. Personally I don't think we loose much by squashing, although it may end up with huge commit diffs in large PRs.

Reply:
> If we are talking about GitHub 'Squash on merge' feature it simply squashes whole PR into one commit and rebases it on top of master (or whatever branch you are merging to). As far as I remember (we use it by default in build.snapcraft.io) it creates a commit message that lists all the messages of commits being squashed.

Yes that's right. The lost information I was referring to is the information about which commit message pertains to which part of the diff, which to me seems pretty important (imagine if we disassociated all of git's commit messages from the diffs), although it does depend on whether the person filing the PR cared about keeping an accurate history.

> I believe we had a dev catch up discussion about merging vs rebasing vs squashing PRs and the consensus was to use merge by default not to loose commit history by squashing whole PR.

Yes I remember that discussion. I agree my memory is that the preference of the team to use merge, which is also my preference. I'm not sure we made a very strong insistence though. We could consider this again, or we could continue with that standard.

Reply:
@squidsoup @bartaz @nottrobin , Should I leave this sentence in here?  The squashing vs merging probably fits better into the github workflow/Pull Request document?

Reply:
@squidsoup since this sentence isn't actually related to merging PRs, AFAICT it's only about keeping the commits in your feature branch  tidy, are you happy with its current form? Or would you like to suggest an alternative?

Reply:
> @squidsoup Or would you like to suggest an alternative?

I personally don't feel there's much to be gained by investing energy in making PRs with commits per feature in large branches, particularly when our workflow for two of our large projects, CRBS and MAAS squash commits on merge. History on master should be clean, but I would rather spend time on the code rather than crafting feature commits when those get lost anyway.

Reply:
> > If we are talking about GitHub 'Squash on merge' feature it simply squashes whole PR into one commit and rebases it on top of master (or whatever branch you are merging to). As far as I remember (we use it by default in build.snapcraft.io) it creates a commit message that lists all the messages of commits being squashed.
> 
Github may do that, but LP does not. Manually writing a multiline commit message for large branches isn't difficult however.

> Yes that's right. The lost information I was referring to is the information about which commit message pertains to which part of the diff, 

I would suggest that if you have to break up your code review by commit, the author should be making three PRs and the branch is too large.

> > I believe we had a dev catch up discussion about merging vs rebasing vs squashing PRs and the consensus was to use merge by default not to loose commit history by squashing whole PR.

Why is intermediary history valuable? Personally speaking, most of my intermediary commits are junk, and little more than placemarkers.

Reply:
> @squidsoup: I personally don't feel there's much to be gained by investing energy in making PRs with commits per feature in large branches, particularly when our workflow for two of our large projects, CRBS and MAAS squash commits on merge. History on master should be clean, but I would rather spend time on the code rather than crafting feature commits when those get lost anyway.

> Why is intermediary history valuable, bearing in mind our workflow on LP? Personally speaking, most of my intermediary commits are junk, and little more than placemarkers.

I completely agree, you should make a call about what makes sense for working on CRBS and MAAS. For almost all of us, this isn't "intermediary history", it's simply "history". But I can see that for you it's different, and it is definitely a waste of time to worry about history that's going to be thrown away.

Firstly, in my opinion this document should only be considered advisory (as I've mentioned a few times) and so you or anyone else should be absolutely free to ignore any of its advice that doesn't work for you.

I assume MAAS and CRBS work this way because we don't control those codebases? So we can assume that we'd try to follow these guidelines for codebases we *do* control? @b-m-f it would be worth, I think, explicitly adding a concession for when we're working on codebases that we don't control.

> I would suggest that if you have to break up your code review by commit, the author should be making three PRs and the branch is too large.

This I completely disagree with. One *could* consider that Git history should merely be a history of all the merged PRs, but this offers *much* less detail than I'm used to working with. PRs should be as small as possible, but at the very least a PR is likely to include several things - e.g. some code, an update to a README, and some tests. I think it's much more useful, if you can, to see PRs as encapsulating whole new pieces of functionality, and commits as describing individual changes.

This inevitably gives you much more detailed history & the commits will have much smaller diffs. If you've ever tried working through history where people have created huge commits I think you'll appreciate how frustrating it can be.

E.g. my PR might have:

```
102: Document user logic
101: Test user logic
100: Improve user logic
```

If I "merge" this PR into master, then my master history will look like:

```
103: Merge PR #56: New user logic (merge of commit 102 & commit 99)
- 102: Document user logic
- 101: Test user logic
- 100: Improve user logic
99: Merge PR #55: Add secret cryptocurrency miner
```

And in GitHub those commits will even tell you that they came from that PR.

If there's no other choice, simply one commit per PR is acceptable, but we can have more detail, I think that's always better.

Reply:
> I assume MAAS and CRBS work this way because we don't control those codebases?

Right, and potentially out of our hands, that's just the way landings work for MAAS and CRBS (RBAC) and I think it's probably best if we try to conform to their workflows as much as possible. I think @b-m-f's suggestions are great, but we probably just need a caveat that they only apply to projects that we have complete control over landing and ci.

Reply:
So, this has been lying around quite a while. 

Do you think that the sentence 
```
But make sure that this is a something that is done throughout your project. If another workflow is used, such as squashing all your commits together before you merge your branch into master this is not necessary. 
```
addresses this concern?

Reply:
@b-m-f sure, that's fine.

Reply:
Yep sounds good to me.